src/lib/tuning/index/base_index_evaluator.cpp

```diff
@@ -0,0 +1,222 @@
+            access_record.condition = predicate_node->predicate_condition();
+            access_record.compare_value = boost::get<AllTypeVariant>(predicate_node->value());
+            _access_records.push_back(access_record);
+          }
+        }
+      } break;
+      case LQPNodeType::Join:
+        // Probably interesting, room for future work
+        break;
+      default:
+        // Not interesting
+        break;
+    }
+  }
+}
+
+void BaseIndexEvaluator::_inspect_pqp_operator(const std::shared_ptr<const AbstractOperator>& op,
+                                               size_t query_frequency) {
+  std::list<const std::shared_ptr<const AbstractOperator>> queue;
+  queue.push_back(op);
+  while (!queue.empty()) {
+    auto node = queue.front();
+    queue.pop_front();
+    if (const auto& table_scan = std::dynamic_pointer_cast<const TableScan>(node)) {
+      /*
+         * A TableScan represents a scan that could possibly be sped up by
+         * creating an index. To do this, we need to find out where this column
+         * came from, i.e. what column on what table is scanned.
+         * Usually the underlying node is a GetTable node, holding this information.
+         */
+      DebugAssert(std::dynamic_pointer_cast<const Validate>(table_scan->input_left()) == nullptr,
+                  "Validation nodes are not supported. Please run the pipeline without MVCC columns.");
+      if (const auto& get_table = std::dynamic_pointer_cast<const GetTable>(table_scan->input_left())) {
+        const auto& table_name = get_table->table_name();
+        ColumnID column_id = table_scan->left_column_id();
+        _access_records.emplace_back(table_name, column_id, query_frequency);
+        _access_records.back().condition = table_scan->predicate_condition();
+        _access_records.back().compare_value = boost::get<AllTypeVariant>(table_scan->right_parameter());
+      }
+    } else {
+      if (node->input_left()) {
+        queue.push_back(node->input_left());
+      }
+      if (node->input_right()) {
+        queue.push_back(node->input_right());
+      }
+    }
+  }
+}
+
+void BaseIndexEvaluator::_aggregate_access_records() {
+  _new_indexes.clear();
+  for (const auto& access_record : _access_records) {
+    _new_indexes.insert(access_record.column_ref);
+    _process_access_record(access_record);
+  }
+}
+
+void BaseIndexEvaluator::_add_existing_indexes() {
+  for (const auto& table_name : StorageManager::get().table_names()) {
+    const auto& table = StorageManager::get().get_table(table_name);
+    const auto& first_chunk = table->get_chunk(ChunkID{0});
+
+    for (const auto& index_info : table->get_indexes()) {
+      auto index_choice = IndexChoice{ColumnRef{table_name, index_info.column_ids}, true};
+      index_choice.type = index_info.type;
+      _choices.emplace_back(index_choice);
+      _new_indexes.erase({table_name, index_info.column_ids});
+    }
+  }
+}
+
+void BaseIndexEvaluator::_add_new_indexes() {
```

Comment:
Here, it would be better to simply inline those three lines instead of having a non-descript method with side effects that you cannot see from the outside. The reader has to look at this anyway to understand what is happening, so why do you make him scroll down this far?